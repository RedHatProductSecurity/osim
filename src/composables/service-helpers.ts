import { getDisplayedOsidbError, parseOsidbErrors } from '@/services/osidb-errors-helpers';
import { useToastStore } from '@/stores/ToastStore';


export function createCatchHandler(title: string = 'Error', shouldThrow: boolean = true) {
  return (error: any) => {
    const displayedError = Array.isArray(error)
      ? parseOsidbErrors(error)
      : getDisplayedOsidbError(error);
    const { addToast } = useToastStore();
    addToast({
      title,
      body: displayedError,
      css: 'warning',
    });
    console.error('‚ùå ', error);

    if (shouldThrow) {
      throw error;
    }
  };
}

export function createSuccessHandler({
  title = 'Operation Successful',
  body,
}: Record<string, string>) {
  return (response: { data: any }) => {
    const { addToast } = useToastStore();
    addToast({
      title,
      body,
      css: 'success',
    });
    return response;
  };
}
