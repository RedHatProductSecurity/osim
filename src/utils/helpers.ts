// Source: https://github.com/vuejs/core/issues/5303#issuecomment-1543596383

import { toRaw, isRef, isReactive, isProxy, ref, toRef, watch } from 'vue';

export function vRef(prop: Record<string, any>, property: string, defaultValue: any) {
  const reffedProp = toRef(prop, property);
  watch(reffedProp, (value) => flexRef.value = value);
  const flexRef = reffedProp.value === undefined ? ref(defaultValue) : reffedProp;
  return flexRef;
}

export function deepToRaw<T extends Record<string, any>>(sourceObj: T): T {
  const objectIterator = (input: any): any => {
    if (Array.isArray(input)) {
      return input.map((item) => objectIterator(item));
    }
    if (isRef(input) || isReactive(input) || isProxy(input)) {
      return objectIterator(toRaw(input));
    }
    if (input && typeof input === 'object') {
      return Object.keys(input).reduce((acc, key) => {
        acc[key as keyof typeof acc] = objectIterator(input[key]);
        return acc;
      }, {} as T);
    }
    return input;
  };

  return objectIterator(sourceObj);
}

export function deepCopyFromRaw<T extends Record<string, any>>(sourceObj: T): T {
  return JSON.parse(JSON.stringify(deepToRaw(sourceObj)));
}

// eslint-disable-next-line max-len
// https://stackoverflow.com/questions/14446511/most-efficient-method-to-groupby-on-an-array-of-objects
export const groupBy = <T>(
  array: T[],
  predicate: (value: T, index: number, array: T[]) => string,
) =>
    array.reduce(
      (acc, value, index, array) => {
        (acc[predicate(value, index, array)] ||= []).push(value);
        return acc;
      },
    {} as { [key: string]: T[] },
    );

export const assignKeyValue = (object: Record<string, any>, key: string, value: any = null) => {
  object[key] = value;
  return object;
};

export const objectMap = (object: Record<string, any>, mapFn: (key: string, value: any) => any) =>
  Object.keys(object).reduce(
    (acc, key) => assignKeyValue(acc, key, mapFn(key, object[key])),
    {} as Record<string, any>,
  );

export const sortedByGroup = <T extends Record<string, any>>(array: T[], key: string) =>
  groupBy(
    array
      .filter((item: T) => item[key])
      .sort((itemA: T, itemB: T) => itemA[key].localeCompare(itemB[key])),
    (item: T) => item[key],
  );
