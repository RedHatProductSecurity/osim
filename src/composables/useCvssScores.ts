import { computed, ref, watch } from 'vue';
import type { ComputedRef } from 'vue';

import { equals, groupWith } from 'ramda';
import { z } from 'zod';

import { useFlawAffectsModel } from '@/composables/useFlawAffectsModel';
import { useFlaw } from '@/composables/useFlaw';
import { useCvss4Calculator } from '@/composables/useCvss4Calculator';
import {
  parseCvss3Factors,
  calculateCvss3Score,
  formatCvss3Factors,
  useCvss3Calculator,
} from '@/composables/useCvss3Calculator';

import { matcherForAffect, deepCopyFromRaw } from '@/utils/helpers';
import { CVSS40 } from '@/utils/cvss40';
import { CvssVersionDisplayMap, CvssVersions, DEFAULT_CVSS_VERSION, CorrespondingCvssFactors } from '@/constants';
import { IssuerEnum } from '@/generated-client';
import { deleteFlawCvssScores, postFlawCvssScores, putFlawCvssScores } from '@/services/FlawService';
import type {
  Dict,
  Nullable,
  ZodAffectType,
  ZodAffectCVSSType,
  ZodFlawCVSSType,
  CvssEntity,
  Cvss,
  ZodFlawType,
} from '@/types';

const wasFlawCvssModified = ref(false);
const shouldSyncVectors = ref(false);
const cvssVersion = ref<CvssVersions>(DEFAULT_CVSS_VERSION);

function filterCvssData(issuer: string, version: string) {
  return (cvss: Cvss) => (cvss.issuer === issuer && cvss.cvss_version === version);
}

function getCvssData(entity: CvssEntity, issuer: string, version: string): Cvss | undefined {
  return entity.cvss_scores?.find(filterCvssData(issuer, version));
}

export function newAffectCvss(cvssVersion?: CvssVersions) {
  const { flaw } = useFlaw();

  return {
    // affect: z.string().uuid(),
    score: null,
    vector: null,
    comment: '',
    cvss_version: cvssVersion ?? DEFAULT_CVSS_VERSION, // TODO: change when affect CVSS4 is supported
    issuer: IssuerEnum.Rh,
    embargoed: flaw.value.embargoed,
    alerts: [],
  } as ZodAffectCVSSType;
}

function newFlawCvss(version: string = DEFAULT_CVSS_VERSION) {
  const { flaw } = useFlaw();

  return {
    score: null,
    vector: null,
    comment: '',
    cvss_version: version,
    issuer: IssuerEnum.Rh,
    embargoed: flaw.value.embargoed,
    created_dt: null,
    uuid: null,
  } as ZodFlawCVSSType;
}

export function validateCvssVector(cvssVector: null | string | undefined, version: CvssVersions) {
  if (version === CvssVersions.V3) {
    const cvss3VectorSchema = z.union([
      z.string().length(44, { message: 'Incomplete CVSS3.1 Vector. There are factors missing.' }),
      z.string().length(0).nullable(),
    ]);
    const parseResult = cvss3VectorSchema.safeParse(cvssVector);
    if (parseResult.success === false) {
      return parseResult.error.errors[0].message;
    }
  }

  if (version === CvssVersions.V4) {
    return new CVSS40(cvssVector).error || null;
  }

  return null;
}

function isAffect(maybeAffect: CvssEntity): maybeAffect is ZodAffectType {
  return 'ps_component' in maybeAffect;
}

function formatScore(score: Nullable<number>) {
  return score?.toFixed(1) ?? '';
}

export function useCvssScores(cvssEntity?: CvssEntity) {
  const { flaw } = useFlaw();

  const rhFlawCvssScores = computed(() => Object.fromEntries(
    Object.values(CvssVersions).map(version => [
      version,
      getCvssData(flaw.value, IssuerEnum.Rh, version) as ZodFlawCVSSType ?? newFlawCvss(version),
    ]),
  ));
  const { updateAffectCvss } = useFlawAffectsModel();
  const { cvss4Score, cvss4Selections, cvss4Vector, errorV4, parseVectorV4String, setMetric } = useCvss4Calculator();
  const wasCvssModified = ref(false);
  const rhCvssScores = ref(rhCvssByVersion());

  function rhCvssByVersion(): Dict<string, Cvss> {
    const entity: CvssEntity = cvssEntity ?? flaw.value;

    for (const version of Object.values(CvssVersions)) {
      if (getCvssData(entity, IssuerEnum.Rh, version)) continue;
      // Supply blank-slate CVSS objects if no scores exist
      if (isAffect(entity)) {
        (entity as ZodAffectType).cvss_scores?.push(newAffectCvss(version));
      } else {
        (entity as ZodFlawType).cvss_scores?.push(newFlawCvss(version));
      }
    }

    const object = Object.fromEntries(
      Object.values(CvssVersions).map((version) => {
        const cvss = getCvssData(entity, IssuerEnum.Rh, version);
        return [version, cvss as Cvss];
      }));
    return object;
  }

  const entity: CvssEntity = cvssEntity ?? flaw.value;
  const flawOrAffect = cvssEntity ? computed(() => cvssEntity) : flaw;

  const flawRhCvss = computed<ZodFlawCVSSType>(() => rhFlawCvssScores.value[cvssVersion.value]);

  const rhCvss = computed<Cvss>(() => rhCvssScores.value[
    // Tentatively, we are only supporting CVSS3 for affects; not clear if per-affect CVSS overrides are used widely
    isAffect(entity) ? CvssVersions.V3 : cvssVersion.value
  ]);

  const matchAffect = matcherForAffect(entity as ZodAffectType);
  const maybeAffect = flaw.value.affects.find(matchAffect);
  const maybeCvss = maybeAffect?.cvss_scores.find(filterCvssData(IssuerEnum.Rh, cvssVersion.value));

  const cvssVector = computed(() => rhCvss.value?.vector ? rhCvss.value?.vector : null);
  const cvssScore = computed(() => rhCvss.value?.score ?? null);
  const { cvss3Factors } = useCvss3Calculator(computed(() => rhCvssScores.value[CvssVersions.V3]?.vector ?? ''));

  const error = computed(() => {
    const errors = [validateCvssVector(cvssVector.value, cvssVersion.value)];
    if (cvssVersion.value === CvssVersions.V4) errors.push(errorV4.value);
    return errors.filter(Boolean).join('. ') || null;
  });

  if (isAffect(entity) && maybeAffect && !maybeCvss) {
    maybeAffect.cvss_scores.push(newAffectCvss());
  }

  const initialRhCvss = deepCopyFromRaw(rhCvss.value);

  watch(rhCvss, () => {
    wasCvssModified.value = !equals(initialRhCvss, rhCvss.value);
    if (!wasCvssModified.value) {
      if (!isAffect(entity)) {
        wasFlawCvssModified.value = false;
      };
      return;
    }
    if (!isAffect(entity)) {
      wasFlawCvssModified.value = true;
      flawRhCvss.value.score = rhCvss.value.score ?? null;
      flawRhCvss.value.vector = rhCvss.value.vector ?? null;
    } else {
      updateAffectCvss(
        entity,
        rhCvss.value.vector ?? '',
        rhCvss.value.score ?? null,
        entity.cvss_scores.findIndex(cvss => cvss.uuid == maybeCvss?.uuid),
      );
    }
    if (shouldSyncVectors.value) synchronizeFactors();
  }, { deep: true });

  watch(() => flaw.value.updated_dt, () => {
    wasFlawCvssModified.value = false;
  });

  watch(cvss3Factors, () => {
    updateUsingV3Vector(formatCvss3Factors(cvss3Factors.value));
  }, { deep: true });

  watch(() => flawOrAffect.value.updated_dt, () => {
    rhCvssScores.value = rhCvssByVersion();
    wasCvssModified.value = false;
  });

  watch(cvss4Score, (score) => {
    if (cvssVersion.value === CvssVersions.V4) updateScore(score);
  });

  watch(cvss4Vector, (vector) => {
    if (cvssVersion.value === CvssVersions.V4) updateVector(vector);
  });

  function synchronizeFactors() {
    const vector = rhCvss.value.vector || '';

    // Convert from CVSS4 to CVSS3
    if (cvssVersion.value === CvssVersions.V4) {
      const cvss4Metrics = cvss4Selections.value.BASE;
      for (const [metric, v4Value] of Object.entries(cvss4Metrics)) {
        if (metric === 'CVSS') continue;
        const factor: string = CorrespondingCvssFactors[metric] ?? metric;

        if (metric === 'UI' && v4Value !== null) {
          // CVSS3 just has N or R for UI metrics, where as CVSS4 has A, P, N for UI metrics
          cvss3Factors.value[factor] = v4Value === 'N' ? 'N' : 'R';
          continue;
        }

        if (factor in cvss3Factors.value) {
          if (cvssScore.value !== null && v4Value === null) continue;
          cvss3Factors.value[factor] = v4Value as string;
          rhCvssScores.value[CvssVersions.V3].vector = formatCvss3Factors({
            ...cvss3Factors.value, CVSS: CvssVersionDisplayMap[CvssVersions.V3],
          });
          rhCvssScores.value[CvssVersions.V3].score = calculateCvss3Score(cvss3Factors.value);
        }
      }
    }

    if (cvssVersion.value === CvssVersions.V3) {
      const factors = parseCvss3Factors(vector);
      for (const [factor, value] of Object.entries(factors)) {
        if (factor === 'CVSS' || value === '') continue;
        const correspondingMetric: string = CorrespondingCvssFactors[factor];
        if (!correspondingMetric) continue;

        // CVSS4 has 2 metrics A, P, N for UI metrics, and only a value of N can be mapped to its corresponding
        // metric factor in CVSS3; A and P would be correct translations of R, but this must be chosen by the user
        const factorValue = (factor === 'UI' && value === 'R') ? null : value;

        cvss4Selections.value['BASE'][correspondingMetric] = factorValue;
      }
      rhCvssScores.value[CvssVersions.V4].vector = cvss4Vector.value;
      rhCvssScores.value[CvssVersions.V4].score = cvss4Score.value;
    }
  }

  async function saveCvssScores() {
    const queue = [];
    for (const cvssData of Object.values(rhFlawCvssScores.value)) {
      // Update logic, if the CVSS score already exists in OSIDB
      if (cvssData.created_dt) {
        // Handle existing CVSS score
        if (cvssData.vector === null && cvssData.uuid != null) {
          queue.push(deleteFlawCvssScores(flaw.value.uuid, cvssData.uuid));
        } else if (cvssData.vector !== null && cvssData.uuid) {
          // Update embargoed state from parent flaw
          cvssData.embargoed = flaw.value.embargoed;
          queue.push(putFlawCvssScores(flaw.value.uuid, cvssData.uuid, cvssData));
        }
        // Handle new CVSS score, since it does not exist in OSIDB yet
      } else if (cvssData.vector !== null) {
        const requestBody = {
          // "score":  is recalculated based on the vector by OSIDB and does not need to be included
          comment: cvssData.comment,
          cvss_version: cvssData.cvss_version,
          issuer: IssuerEnum.Rh,
          vector: cvssData.vector,
          embargoed: flaw.value.embargoed,
        };
        queue.push(postFlawCvssScores(flaw.value.uuid, requestBody));
      }
      // Handle newly created CVSS score
    }

    return Promise.all(queue);
  }

  function updateScore(score: null | number) {
    rhCvss.value.score = score;
  }

  function updateVector(vector: null | string) {
    rhCvss.value.vector = vector;
    if (cvssVersion.value === CvssVersions.V4 && vector === null) parseVectorV4String(null);
    if (cvssVersion.value === CvssVersions.V3 && vector === null) cvss3Factors.value = {};
  }

  function updateUsingV3Vector(newCvssVector: null | string | undefined) {
    // Should this just update V3 instead of returnin if v4
    if (cvssVersion.value === CvssVersions.V4) return;

    updateVector(newCvssVector ?? '');
    const factors = parseCvss3Factors(newCvssVector ?? '');
    updateScore(calculateCvss3Score(factors) ?? 0);
  }

  function reset() {
    updateVector(null);
    updateScore(null);
  }

  return {
    updateVector,
    updateUsingV3Vector,
    updateScore,
    cvss3Factors,
    cvssVersion,
    cvssVector,
    cvss4Score,
    cvss4Selections,
    cvss4Vector,
    cvssScore,
    wasCvssModified,
    wasFlawCvssModified,
    flawRhCvss,
    rhCvss,
    reset,
    parseVectorV4String,
    saveCvssScores,
    setMetric,
    shouldSyncVectors,
    ...useFlawCvssStrings(flawRhCvss),
    error,
  };
}

function useFlawCvssStrings(flawRhCvss: ComputedRef<ZodFlawCVSSType>) {
  const { flaw } = useFlaw();

  const flawNvdCvss = computed(() => getCvssData(flaw.value, IssuerEnum.Nist, cvssVersion.value));

  const nvdCvssString = computed(() => {
    const values = [formatScore(flawNvdCvss.value?.score), flawNvdCvss.value?.vector].filter(Boolean);
    return values.join(' ') || '-';
  });

  const rhCvssString = computed(() => {
    if (flawRhCvss.value.cvss_version === CvssVersions.V3) {
      const values = [formatScore(flawRhCvss.value?.score), flawRhCvss.value?.vector].filter(Boolean);
      return values.join(' ');
    }
    if (flawRhCvss.value.cvss_version === CvssVersions.V4) {
      const { score, vector } = flawRhCvss.value;
      return score + ' ' + vector;
    }
    return '-';
  });

  const shouldDisplayEmailNistForm = computed(() => {
    if (['', '-'].includes(rhCvssString.value) || ['', '-'].includes(nvdCvssString.value)) {
      return false;
    }
    if (flawRhCvss.value.comment) {
      return true;
    }
    return `${rhCvssString.value}` !== `${nvdCvssString.value}`;
  });

  const highlightedNvdCvssString = computed(() => {
    if (!flawNvdCvss.value?.vector
      || flawNvdCvss.value?.vector === '-'
      || !flawRhCvss.value?.vector) {
      return [[{ char: '-', isHighlighted: false }]];
    }

    const result = [];
    const nvdCvssValue = flawNvdCvss.value?.vector || '-';
    const cvssValue = flawRhCvss.value?.vector || '-';
    const maxLength = Math.max(nvdCvssValue.length, cvssValue.length);

    if (formatScore(flawNvdCvss.value?.score) !== formatScore(flawRhCvss.value?.score)) {
      result.push(
        { char: formatScore(flawNvdCvss.value?.score), isHighlighted: true },
        { char: ' ', isHighlighted: false },
      );
    }

    for (let i = 0; i < maxLength; i++) {
      const charFromFlaw = i < nvdCvssValue.length ? nvdCvssValue[i] : '';
      const charFromCvss = i < cvssValue.length ? cvssValue[i] : '';
      result.push({
        char: charFromFlaw,
        isHighlighted: shouldDisplayEmailNistForm.value && charFromFlaw !== charFromCvss,
      });
    }

    return groupWith((a, b) => a.isHighlighted === b.isHighlighted, result);
  });

  return {
    rhCvssString,
    nvdCvssString,
    highlightedNvdCvssString,
    shouldDisplayEmailNistForm,
  };
}
