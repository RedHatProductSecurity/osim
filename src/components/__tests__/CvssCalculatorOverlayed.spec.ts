import { type Component } from 'vue';

import sampleFlawFull from '@test-fixtures/sampleFlawFull.json';
import { VueWrapper, mount } from '@vue/test-utils';
import { describe, it, expect } from 'vitest';

import { useCvssScores, validateCvssVector } from '@/composables/useCvssScores';

import { importActual } from '@/__tests__/helpers';

vi.mock('@/composables/useCvssScores');

async function activateFactorButton(subject: any, factor: string, value: string) {
  const factorButtonGroup = subject.findAll('.btn-group-vertical')
    .filter((node: { text: () => string | string[] }) => node.text().includes(factor))[0];
  const factorButton = factorButtonGroup
    .findAll('.btn')
    .filter((node: { text: () => string }) => node.text() === value)[0];
  await factorButton.trigger('click');
}

async function activateMultipleFactorButtons(subject: any, factorValuePairs: [string, string][]) {
  for (const [factor, value] of factorValuePairs) {
    await activateFactorButton(subject, factor, value);
  }
}

describe('cvssCalculatorOverlayed', () => {
  type Subject = VueWrapper<Component>;
  let subject: Subject;
  beforeEach(async () => {
    const {
      useCvssScores: _useCvssScores,
      validateCvssVector: _validateCvssVector,
    } = await importActual('@/composables/useCvssScores');
    vi.mocked(useCvssScores).mockImplementation(_useCvssScores);
    vi.mocked(validateCvssVector).mockImplementation(_validateCvssVector);

    const importedComponent = await import('@/components/CvssCalculator/CvssCalculatorOverlayed.vue');
    const CvssCalculatorOverlayed = importedComponent.default;
    subject = mount(CvssCalculatorOverlayed, {
      props: {
        affect: sampleFlawFull.affects[0],
      },
    });
  });

  afterEach(() => {
    vi.resetAllMocks();
    vi.resetModules();
    subject?.unmount();
  });

  it('mounts and renders correctly', async () => {
    await subject.find('.vector-input').trigger('focus');
    expect(subject.html()).toMatchSnapshot();
  });

  it('shows calculator on input focus', async () => {
    await subject.find('.vector-input').trigger('focus');
    expect(subject.find('.cvss-calculator').classes('visually-hidden')).toBe(false);
  });

  it('hides calculator on input blur', async () => {
    await subject.find('.vector-input');
    expect(subject.find('.cvss-calculator').classes('visually-hidden')).toBe(true);
  });

  // Validations
  it('doesn\'t show validation on empty vector', async () => {
    const inputVectorValue = subject.find('.vector-input');
    expect(inputVectorValue.classes().includes('is-invalid')).toBe(false);
  });

  it('doesn\'t show validation on complete vector', async () => {
    const factorValuePairs: [string, string][] = [
      ['Attack Vector', 'Network'],
      ['Attack Complexity', 'Low'],
      ['Privileges Required', 'None'],
      ['User Interaction', 'None'],
      ['Scope', 'Changed'],
      ['Confidentiality', 'High'],
      ['Integrity', 'High'],
      ['Availability', 'High'],
    ];
    await activateMultipleFactorButtons(subject, factorValuePairs);
    const inputVectorValue = subject.find('.vector-input');
    expect(inputVectorValue.classes().includes('is-p')).toBe(false);
  });

  it('shows validation on incomplete vector', async () => {
    const inputVectorValue = subject.find('.vector-input');
    await activateFactorButton(subject, 'Availability', 'High');
    expect(inputVectorValue.classes().includes('is-invalid')).toBe(true);
  });
});
