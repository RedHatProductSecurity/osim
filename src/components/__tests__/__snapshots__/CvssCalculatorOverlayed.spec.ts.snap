// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cvssCalculatorOverlayed > mounts and renders correctly 1`] = `
"<div data-v-b418eb65="" tabindex="0"><span data-v-b418eb65="">0</span><i data-v-b418eb65="" class="bi bi-calculator-fill p-2"></i>
  <div data-v-707feb84="" data-v-b418eb65="" class="cvss-calculator overlayed overlayed" affect="[object Object]" cvssscore="0">
    <div data-v-707feb84="" class="p-3">
      <div data-v-b418eb65="" class="osim-input">
        <div data-v-b418eb65="" class="input-wrapper col">
          <div data-v-1d5c2355="" data-v-b418eb65="" tabindex="0" class="dark-background text-cursor vector-input form-control vector-input vector-input" style="height: 30.8px;" focused=""><span data-v-1d5c2355="" class="osim-cvss-score" style="color: white;">0 </span>
            <!--v-if-->
            <!--v-if-->
          </div>
        </div><button data-v-b418eb65="" tabindex="-1" type="button" disabled="" class="erase-button input-group-text"><i data-v-b418eb65="" class="bi bi-eraser"></i></button>
      </div>
      <div data-v-707feb84="" class="osim-input mx-0">
        <div data-v-707feb84="" class="mx-auto">
          <div data-v-707feb84="" class="row-group">
            <div data-v-707feb84="" class="col-group">
              <div data-v-707feb84="" class="btn-group-vertical btn-group-sm osim-factor-severity-select"><button data-v-707feb84="" class="btn-group-header btn lh-sm" disabled="">Attack Vector</button><button data-v-707feb84="" tabindex="-1" type="button" class="btn lh-sm" data-bs-toggle="tooltip" data-bs-placement="right" title="Worst: The vulnerable component is bound to the network stack and the set of possible attackers extends beyond the other options listed below, up to and including the entire Internet. Such a vulnerability is often termed “remotely exploitable” and can be thought of as an attack being exploitable at the protocol level one or more network hops away (e.g., across one or more routers)." style="background-color: rgb(224, 224, 224);">Network</button><button data-v-707feb84="" tabindex="-1" type="button" class="btn lh-sm" data-bs-toggle="tooltip" data-bs-placement="right" title="Worse: The vulnerable component is bound to the network stack, but the attack is limited at the protocol level to a logically adjacent topology. This can mean an attack must be launched from the same shared physical (e.g., Bluetooth or IEEE 802.11) or logical (e.g., local IP subnet) network, or from within a secure or otherwise limited administrative domain (e.g., MPLS, secure VPN to an administrative network zone). One example of an Adjacent attack would be an ARP (IPv4) or neighbor discovery (IPv6) flood leading to a denial of service on the local LAN segment." style="background-color: rgb(224, 224, 224);">Adjacent</button><button data-v-707feb84="" tabindex="-1" type="button" class="btn lh-sm" data-bs-toggle="tooltip" data-bs-placement="right" title="Bad: The vulnerable component is not bound to the network stack and the attacker’s path is via read/write/execute capabilities. Either: 1.the attacker exploits the vulnerability by accessing the target system locally (e.g., keyboard, console), or remotely (e.g., SSH); 2.or the attacker relies on User Interaction by another person to perform actions required to exploit the vulnerability (e.g., using social engineering techniques to trick a legitimate user into opening a malicious document)." style="background-color: rgb(224, 224, 224);">Local</button><button data-v-707feb84="" tabindex="-1" type="button" class="btn lh-sm" data-bs-toggle="tooltip" data-bs-placement="right" title="Bad: The attack requires the attacker to physically touch or manipulate the vulnerable component. Physical interaction may be brief (e.g., evil maid attack) or persistent. An example of such an attack is a cold boot attack in which an attacker gains access to disk encryption keys after physically accessing the target system. Other examples include peripheral attacks via FireWire/USB Direct Memory Access (DMA)." style="background-color: rgb(224, 224, 224);">Physical</button></div>
            </div>
            <div data-v-707feb84="" class="col-group">
              <div data-v-707feb84="" class="btn-group-vertical btn-group-sm osim-factor-severity-select"><button data-v-707feb84="" class="btn-group-header btn lh-sm" disabled="">Attack Complexity</button><button data-v-707feb84="" tabindex="-1" type="button" class="btn lh-sm" data-bs-toggle="tooltip" data-bs-placement="right" title="Worst: Specialized access conditions or extenuating circumstances do not exist. An attacker can expect repeatable success when attacking the vulnerable component." style="background-color: rgb(224, 224, 224);">Low</button><button data-v-707feb84="" tabindex="-1" type="button" class="btn lh-sm" data-bs-toggle="tooltip" data-bs-placement="right" title="Bad: A successful attack depends on conditions beyond the attacker's control. That is, a successful attack cannot be accomplished at will, but requires the attacker to invest in some measurable amount of effort in preparation or execution against the vulnerable component before a successful attack can be expected." style="background-color: rgb(224, 224, 224);">High</button></div>
            </div>
            <div data-v-707feb84="" class="col-group">
              <div data-v-707feb84="" class="btn-group-vertical btn-group-sm osim-factor-severity-select"><button data-v-707feb84="" class="btn-group-header btn lh-sm" disabled="">Privileges Required</button><button data-v-707feb84="" tabindex="-1" type="button" class="btn lh-sm" data-bs-toggle="tooltip" data-bs-placement="right" title="Worst: The attacker is unauthorized prior to attack, and therefore does not require any access to settings or files of the the vulnerable system to carry out an attack." style="background-color: rgb(224, 224, 224);">None</button><button data-v-707feb84="" tabindex="-1" type="button" class="btn lh-sm" data-bs-toggle="tooltip" data-bs-placement="right" title="Worse: The attacker requires privileges that provide basic user capabilities that could normally affect only settings and files owned by a user. Alternatively, an attacker with Low privileges has the ability to access only non-sensitive resources." style="background-color: rgb(224, 224, 224);">Low</button><button data-v-707feb84="" tabindex="-1" type="button" class="btn lh-sm" data-bs-toggle="tooltip" data-bs-placement="right" title="Bad: The attacker requires privileges that provide significant (e.g., administrative) control over the vulnerable component allowing access to component-wide settings and files." style="background-color: rgb(224, 224, 224);">High</button></div>
            </div>
            <div data-v-707feb84="" class="col-group">
              <div data-v-707feb84="" class="btn-group-vertical btn-group-sm osim-factor-severity-select"><button data-v-707feb84="" class="btn-group-header btn lh-sm" disabled="">User Interaction</button><button data-v-707feb84="" tabindex="-1" type="button" class="btn lh-sm" data-bs-toggle="tooltip" data-bs-placement="right" title="Worst: The vulnerable system can be exploited without interaction from any user." style="background-color: rgb(224, 224, 224);">None</button><button data-v-707feb84="" tabindex="-1" type="button" class="btn lh-sm" data-bs-toggle="tooltip" data-bs-placement="right" title="Bad: Successful exploitation of this vulnerability requires a user to take some action before the vulnerability can be exploited. For example, a successful exploit may only be possible during the installation of an application by a system administrator." style="background-color: rgb(224, 224, 224);">Required</button></div>
            </div>
          </div>
          <div data-v-707feb84="" class="row-group">
            <div data-v-707feb84="" class="col-group">
              <div data-v-707feb84="" class="btn-group-vertical btn-group-sm osim-factor-severity-select"><button data-v-707feb84="" class="btn-group-header btn lh-sm" disabled="">Scope</button><button data-v-707feb84="" tabindex="-1" type="button" class="btn lh-sm" data-bs-toggle="tooltip" data-bs-placement="right" title="Worst: An exploited vulnerability can affect resources beyond the security scope managed by the security authority of the vulnerable component. In this case, the vulnerable component and the impacted component are different and managed by different security authorities." style="background-color: rgb(224, 224, 224);">Changed</button><button data-v-707feb84="" tabindex="-1" type="button" class="btn lh-sm" data-bs-toggle="tooltip" data-bs-placement="right" title="Bad: An exploited vulnerability can only affect resources managed by the same security authority. In this case, the vulnerable component and the impacted component are either the same, or both are managed by the same security authority." style="background-color: rgb(224, 224, 224);">Unchanged</button></div>
            </div>
            <div data-v-707feb84="" class="col-group">
              <div data-v-707feb84="" class="btn-group-vertical btn-group-sm osim-factor-severity-select"><button data-v-707feb84="" class="btn-group-header btn lh-sm" disabled="">Confidentiality</button><button data-v-707feb84="" tabindex="-1" type="button" class="btn lh-sm" data-bs-toggle="tooltip" data-bs-placement="right" title="Worst: There is a total loss of confidentiality, resulting in all resources within the impacted component being divulged to the attacker. Alternatively, access to only some restricted information is obtained, but the disclosed information presents a direct, serious impact. For example, an attacker steals the administrator's password, or private encryption keys of a web server." style="background-color: rgb(224, 224, 224);">High</button><button data-v-707feb84="" tabindex="-1" type="button" class="btn lh-sm" data-bs-toggle="tooltip" data-bs-placement="right" title="Bad: There is some loss of confidentiality. Access to some restricted information is obtained, but the attacker does not have control over what information is obtained, or the amount or kind of loss is limited. The information disclosure does not cause a direct, serious loss to the impacted component." style="background-color: rgb(224, 224, 224);">Low</button><button data-v-707feb84="" tabindex="-1" type="button" class="btn lh-sm" data-bs-toggle="tooltip" data-bs-placement="right" title="Good: There is no loss of confidentiality within the impacted component." style="background-color: rgb(224, 224, 224);">None</button></div>
            </div>
            <div data-v-707feb84="" class="col-group">
              <div data-v-707feb84="" class="btn-group-vertical btn-group-sm osim-factor-severity-select"><button data-v-707feb84="" class="btn-group-header btn lh-sm" disabled="">Integrity</button><button data-v-707feb84="" tabindex="-1" type="button" class="btn lh-sm" data-bs-toggle="tooltip" data-bs-placement="right" title="Worst: There is a total loss of integrity, or a complete loss of protection. For example, the attacker is able to modify any/all files protected by the impacted component. Alternatively, only some files can be modified, but malicious modification would present a direct, serious consequence to the impacted component." style="background-color: rgb(224, 224, 224);">High</button><button data-v-707feb84="" tabindex="-1" type="button" class="btn lh-sm" data-bs-toggle="tooltip" data-bs-placement="right" title="Bad: Modification of data is possible, but the attacker does not have control over the consequence of a modification, or the amount of modification is limited. The data modification does not have a direct, serious impact on the impacted component." style="background-color: rgb(224, 224, 224);">Low</button><button data-v-707feb84="" tabindex="-1" type="button" class="btn lh-sm" data-bs-toggle="tooltip" data-bs-placement="right" title="Good: There is no loss of integrity within the impacted component." style="background-color: rgb(224, 224, 224);">None</button></div>
            </div>
            <div data-v-707feb84="" class="col-group">
              <div data-v-707feb84="" class="btn-group-vertical btn-group-sm osim-factor-severity-select"><button data-v-707feb84="" class="btn-group-header btn lh-sm" disabled="">Availability</button><button data-v-707feb84="" tabindex="-1" type="button" class="btn lh-sm" data-bs-toggle="tooltip" data-bs-placement="right" title="Worst: There is a total loss of availability, resulting in the attacker being able to fully deny access to resources in the impacted component; this loss is either sustained (while the attacker continues to deliver the attack) or persistent (the condition persists even after the attack has completed). Alternatively, the attacker has the ability to deny some availability, but the loss of availability presents a direct, serious consequence to the impacted component (e.g., the attacker cannot disrupt existing connections, but can prevent new connections; the attacker can repeatedly exploit a vulnerability that, in each instance of a successful attack, leaks a only small amount of memory, but after repeated exploitation causes a service to become completely unavailable)." style="background-color: rgb(224, 224, 224);">High</button><button data-v-707feb84="" tabindex="-1" type="button" class="btn lh-sm" data-bs-toggle="tooltip" data-bs-placement="right" title="Bad: Performance is reduced or there are interruptions in resource availability. Even if repeated exploitation of the vulnerability is possible, the attacker does not have the ability to completely deny service to legitimate users. The resources in the impacted component are either partially available all of the time, or fully available only some of the time, but overall there is no direct, serious consequence to the impacted component." style="background-color: rgb(224, 224, 224);">Low</button><button data-v-707feb84="" tabindex="-1" type="button" class="btn lh-sm" data-bs-toggle="tooltip" data-bs-placement="right" title="Good: There is no impact to availability within the impacted component." style="background-color: rgb(224, 224, 224);">None</button></div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;
